File: pkg/client/audit/audit.go
// Package audit provides audit log client services for the remote-signer.
package audit

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
	"strings"
	"time"

	"github.com/ivanzzeth/remote-signer/pkg/client/internal/transport"
)

// Service handles audit log operations.
type Service struct {
	transport *transport.Transport
}

// NewService creates a new audit service.
func NewService(t *transport.Transport) *Service {
	return &Service{transport: t}
}

// List lists audit records with optional filters.
func (s *Service) List(ctx context.Context, filter *ListFilter) (*ListResponse, error) {
	path := "/api/v1/audit"
	params := make([]string, 0)

	if filter != nil {
		if filter.EventType != "" {
			params = append(params, fmt.Sprintf("event_type=%s", filter.EventType))
		}
		if filter.Severity != "" {
			params = append(params, fmt.Sprintf("severity=%s", filter.Severity))
		}
		if filter.APIKeyID != "" {
			params = append(params, fmt.Sprintf("api_key_id=%s", filter.APIKeyID))
		}
		if filter.SignerAddress != "" {
			params = append(params, fmt.Sprintf("signer_address=%s", filter.SignerAddress))
		}
		if filter.ChainType != "" {
			params = append(params, fmt.Sprintf("chain_type=%s", filter.ChainType))
		}
		if filter.ChainID != "" {
			params = append(params, fmt.Sprintf("chain_id=%s", filter.ChainID))
		}
		if filter.StartTime != nil {
			params = append(params, fmt.Sprintf("start_time=%s", filter.StartTime.Format(time.RFC3339)))
		}
		if filter.EndTime != nil {
			params = append(params, fmt.Sprintf("end_time=%s", filter.EndTime.Format(time.RFC3339)))
		}
		if filter.Limit > 0 {
			params = append(params, fmt.Sprintf("limit=%d", filter.Limit))
		}
		if filter.Cursor != nil {
			params = append(params, fmt.Sprintf("cursor=%s", url.QueryEscape(*filter.Cursor)))
		}
		if filter.CursorID != nil {
			params = append(params, fmt.Sprintf("cursor_id=%s", url.QueryEscape(*filter.CursorID)))
		}
	}

	if len(params) > 0 {
		path += "?" + strings.Join(params, "&")
	}

	var listResp ListResponse
	err := s.transport.Request(ctx, http.MethodGet, path, nil, &listResp, http.StatusOK)
	if err != nil {
		return nil, err
	}
	return &listResp, nil
}

// API defines the audit service interface.
type API interface {
	List(ctx context.Context, filter *ListFilter) (*ListResponse, error)
}

// Compile-time interface check.
var _ API = (*Service)(nil)

File: pkg/rs-client/src/audit/audit.rs
use reqwest::Method;

use crate::error::Error;
use crate::transport::transport::Transport;

use super::{ListFilter, ListResponse};

#[derive(Clone)]
pub struct Service {
    transport: Transport,
}

impl Service {
    pub fn new(transport: Transport) -> Self {
        Self { transport }
    }

    pub fn list(&self, filter: Option<&ListFilter>) -> Result<ListResponse, Error> {
        let mut path = String::from("/api/v1/audit");
        let mut params = vec![];
        if let Some(f) = filter {
            if let Some(v) = &f.event_type {
                params.push(format!("event_type={}", urlencoding::encode(v)));
            }
            if let Some(v) = &f.severity {
                params.push(format!("severity={}", urlencoding::encode(v)));
            }
            if let Some(v) = &f.api_key_id {
                params.push(format!("api_key_id={}", urlencoding::encode(v)));
            }
            if let Some(v) = &f.signer_address {
                params.push(format!("signer_address={}", urlencoding::encode(v)));
            }
            if let Some(v) = &f.chain_type {
                params.push(format!("chain_type={}", urlencoding::encode(v)));
            }
            if let Some(v) = &f.chain_id {
                params.push(format!("chain_id={}", urlencoding::encode(v)));
            }
            if let Some(v) = &f.start_time {
                params.push(format!("start_time={}", urlencoding::encode(&v.format(&time::format_description::well_known::Rfc3339).map_err(|e| Error::InvalidConfig(e.to_string()))?)));
            }
            if let Some(v) = &f.end_time {
                params.push(format!("end_time={}", urlencoding::encode(&v.format(&time::format_description::well_known::Rfc3339).map_err(|e| Error::InvalidConfig(e.to_string()))?)));
            }
            if let Some(v) = f.limit {
                params.push(format!("limit={}", v));
            }
            if let Some(v) = &f.cursor {
                params.push(format!("cursor={}", urlencoding::encode(v)));
            }
            if let Some(v) = &f.cursor_id {
                params.push(format!("cursor_id={}", urlencoding::encode(v)));
            }
        }
        if !params.is_empty() {
            path.push('?');
            path.push_str(&params.join("&"));
        }

        self.transport
            .request_json(Method::GET, &path, Option::<&()>::None, Some(&[200]))
    }
}
